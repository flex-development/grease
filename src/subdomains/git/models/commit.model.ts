/**
 * @file Models - Commit
 * @module grease/git/models/Commit
 */

import type { RepositoryProvider } from '#src/git/enums'
import type { ICommit } from '#src/git/interfaces'
import type { CommitGrammarOptions } from '#src/git/options'
import type {
  Author,
  BreakingChange,
  CommitFields,
  Reference,
  Trailer
} from '#src/git/types'
import {
  at,
  cast,
  construct,
  defaults,
  get,
  ifelse,
  objectify,
  pick,
  select,
  split,
  trim,
  trimEnd,
  unique,
  type Nullable,
  type Simplify
} from '@flex-development/tutils'
import CommitGrammar from './commit-grammar.model'
import Repository from './repository.model'

/**
 * Parsed commit model.
 *
 * @class
 * @implements {ICommit}
 */
class Commit implements ICommit {
  /**
   * Raw commit chunk.
   *
   * @protected
   * @readonly
   * @instance
   * @member {string} chunk
   */
  protected readonly chunk: string

  /**
   * Raw commit fields object.
   *
   * @see {@linkcode CommitFields}
   *
   * @protected
   * @readonly
   * @instance
   * @member {CommitFields} fields
   */
  protected readonly fields: CommitFields

  /**
   * Commit grammar instance.
   *
   * @see {@linkcode CommitGrammar}
   *
   * @public
   * @readonly
   * @instance
   * @member {CommitGrammar} grammar
   */
  public readonly grammar: CommitGrammar

  /**
   * Repository provider name.
   *
   * @public
   * @readonly
   * @instance
   * @member {RepositoryProvider} provider
   */
  public readonly provider: RepositoryProvider

  /**
   * Parent repository.
   *
   * @public
   * @readonly
   * @instance
   * @member {Repository} repository
   */
  public readonly repository: Repository

  /**
   * Create a new parsed commit.
   *
   * @param {Buffer | Commit | string} chunk - Raw commit chunk or parsed commit
   * @param {Repository?} [repository=new Repository()] - Parent repository
   * @param {Partial<CommitGrammarOptions>?} [grammar] - Commit grammar options
   */
  constructor(
    chunk: Buffer | Commit | string,
    repository: Repository = new Repository(),
    grammar?: Partial<CommitGrammarOptions>
  ) {
    this.chunk = (chunk = chunk.toString()).padEnd(chunk.length + 1, '\n')
    if (!trim(this.chunk)) throw new Error('empty commit chunk')

    // set repository and provider
    this.repository = get(chunk, 'repository', repository)
    this.provider = this.repository.provider

    // set commit grammar
    this.grammar = get(chunk, 'grammar', new CommitGrammar({
      ...grammar,
      provider: this.provider
    }))

    // init commit fields object
    this.fields = construct(cast<CommitFields>(objectify(
      [...this.chunk.matchAll(this.grammar.field)],
      match => get(match, 'groups.field', ''),
      match => trim(get(match, 'groups.value', ''))
    )))

    // add header data to commit fields object
    this.fields = defaults(
      this.fields,
      pick(get(this.grammar.header.exec(this.fields.header), 'groups')!, [
        'breaking',
        'pr',
        'scope',
        'subject',
        'type'
      ]),
      {
        breaking: '',
        pr: 'null',
        scope: '',
        subject: '',
        type: ''
      }
    )

    // remove extra chunk padding
    this.chunk = this.chunk.slice(0, -1)
  }

  /**
   * Commit author details.
   *
   * @see {@linkcode Author}
   *
   * @public
   *
   * @return {Author} Commit author object
   */
  public get author(): Author {
    return this.fields.author
  }

  /**
   * Commit body text.
   *
   * @public
   *
   * @return {string} Commit body
   */
  public get body(): string {
    return trimEnd(this.fields.body.replace(this.fields.trailers, ''))
  }

  /**
   * Commit contains breaking changes?
   *
   * @public
   *
   * @return {boolean} Boolean indicating commits contains breaking changes
   */
  public get breaking(): boolean {
    return this.breaks.length > 0
  }

  /**
   * Breaking changes noted in {@linkcode subject} and {@linkcode trailers}.
   *
   * @see {@linkcode BreakingChange}
   *
   * @public
   *
   * @return {BreakingChange<Commit>[]} Breaking changes array
   */
  public get breaks(): BreakingChange<Commit>[] {
    return unique(select(
      [
        this.fields.breaking
          ? {
            pr: this.pr,
            scope: this.scope,
            token: 'BREAKING-CHANGE',
            type: this.type,
            value: this.subject
          }
          : <Trailer>{ token: this.fields.breaking },
        ...this.trailers
      ],
      trailer => /^BREAKING[ -]CHANGE/.test(trailer.token),
      trailer => ({
        commit: this,
        mentions: this.mentions,
        pr: get(trailer, 'pr', null),
        scope: get(trailer, 'scope', null),
        subject: trailer.value,
        type: this.type
      })
    ))
  }

  /**
   * Commit date in strict ISO 8601 format (`%cI`).
   *
   * @see https://git-scm.com/docs/pretty-formats/2.21.0
   *
   * @public
   *
   * @return {string} Commit date
   */
  public get date(): string {
    return this.fields.date
  }

  /**
   * Abbreviated commit {@linkcode sha}.
   *
   * @public
   *
   * @return {string} Commit hash
   */
  public get hash(): string {
    return this.fields.hash
  }

  /**
   * Commit {@linkcode type}, {@linkcode scope}, breaking change indicator, and
   * {@linkcode subject}.
   *
   * @public
   *
   * @return {string} Commit header text
   */
  public get header(): string {
    return this.fields.header
  }

  /**
   * Users mentioned in commit.
   *
   * @public
   *
   * @return {string[]} Mentions array
   */
  public get mentions(): string[] {
    return select(
      [...this.chunk.matchAll(this.grammar.mention)],
      null,
      match => get(match, 'groups.user')!
    )
  }

  /**
   * Pull request number.
   *
   * @public
   *
   * @return {Nullable<number>} Pull request number or `null`
   */
  public get pr(): Nullable<number> {
    return cast(JSON.parse(this.fields.pr))
  }

  /**
   * Issue and pull request references.
   *
   * @see {@linkcode Reference}
   *
   * @public
   *
   * @return {Reference[]} Reference array
   */
  public get references(): Reference[] {
    return select(
      [...this.chunk.matchAll(this.grammar.reference)],
      null,
      m => ({
        action: get(m, 'groups.action', null),
        number: JSON.parse(get(m, 'groups.number', 'null')),
        owner: get(m, 'groups.owner', null),
        ref: get(m, 'groups.ref', ''),
        repo: get(m, 'groups.repo', null)
      })
    )
  }

  /**
   * Commit scope.
   *
   * @public
   *
   * @return {Nullable<string>} Commit scope or `null`
   */
  public get scope(): Nullable<string> {
    return ifelse(this.fields.scope, this.fields.scope, null)
  }

  /**
   * Commit sha.
   *
   * @public
   *
   * @return {string} Commit sha
   */
  public get sha(): string {
    return this.fields.sha
  }

  /**
   * Commit subject.
   *
   * @public
   *
   * @return {string} Commit subject text
   */
  public get subject(): string {
    return this.fields.subject
  }

  /**
   * Tags associated with commit.
   *
   * @public
   *
   * @return {string[]} Tags array
   */
  public get tags(): string[] {
    return select(
      select(
        split(this.fields.tags, ','),
        tag => !!tag.length,
        tag => trim(tag.replace(/^tag: */, ''))
      ),
      tag => this.grammar.tag.test(tag)
    )
  }

  /**
   * Commit message trailers.
   *
   * @public
   *
   * @return {Trailer[]} Trailers array
   */
  public get trailers(): Trailer[] {
    return select(
      [...this.fields.trailers.matchAll(this.grammar.trailer)],
      null,
      match => pick(match.groups!, ['token', 'value'])
    )
  }

  /**
   * Commit type.
   *
   * @public
   *
   * @return {string} Commit type
   */
  public get type(): string {
    return this.fields.type
  }

  /**
   * Tagged commit version.
   *
   * @public
   *
   * @return {Nullable<string>} First tag in {@linkcode tags} or `null`
   */
  public get version(): Nullable<string> {
    return at(this.tags, 0, null)
  }

  /**
   * Get a JSON-serializable commit object.
   *
   * @public
   *
   * @return {Simplify<ICommit>} JSON-serializable commit
   */
  public toJSON(): Simplify<ICommit> {
    return {
      author: this.author,
      body: this.body,
      breaking: this.breaking,
      breaks: this.breaks,
      date: this.date,
      hash: this.hash,
      header: this.header,
      mentions: this.mentions,
      pr: this.pr,
      references: this.references,
      scope: this.scope,
      sha: this.sha,
      subject: this.subject,
      tags: this.tags,
      trailers: this.trailers,
      type: this.type,
      version: this.version
    }
  }

  /**
   * Get `this` raw commit.
   *
   * @public
   *
   * @return {string} Raw commit string
   */
  public toString(): string {
    return this.chunk
  }
}

export default Commit
