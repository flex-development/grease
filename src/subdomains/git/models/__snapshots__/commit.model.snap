// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:git/models/Commit > #body > should return empty commit body 1`] = `""`;

exports[`unit:git/models/Commit > #body > should return non-empty commit body 1`] = `"P010-28"`;

exports[`unit:git/models/Commit > #breaks > should return empty breaking changes array 1`] = `[]`;

exports[`unit:git/models/Commit > #breaks > should return non-empty breaking changes array 1`] = `
[
  {
    "commit": {
      "author": {
        "email": "unicornware@flexdevelopment.llc",
        "name": "Lexus Drumgold",
      },
      "body": "",
      "breaking": true,
      "breaks": [
        {
          "commit": [Circular],
          "mentions": [],
          "pr": null,
          "scope": null,
          "subject": "init project rewrite",
          "type": "refactor",
        },
      ],
      "date": "2023-09-04T16:26:30-04:00",
      "hash": "455c149",
      "header": "refactor!: init project rewrite",
      "mentions": [],
      "pr": null,
      "references": [],
      "scope": null,
      "sha": "455c1497e3ed5019ab0ebd234509703dadba206a",
      "subject": "init project rewrite",
      "tags": [],
      "trailers": [
        {
          "token": "Signed-off-by",
          "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
        },
      ],
      "type": "refactor",
      "version": null,
    },
    "mentions": [],
    "pr": null,
    "scope": null,
    "subject": "init project rewrite",
    "type": "refactor",
  },
]
`;

exports[`unit:git/models/Commit > #date > should return commit date 1`] = `"2021-09-18T23:29:18-04:00"`;

exports[`unit:git/models/Commit > #hash > should return abbreviated #sha 1`] = `"e003836"`;

exports[`unit:git/models/Commit > #header > should return commit header with #grammar.header match 1`] = `"feat(cli): \`BumpCommand\`"`;

exports[`unit:git/models/Commit > #header > should return commit header without #grammar.header match 1`] = `"initial commit"`;

exports[`unit:git/models/Commit > #mentions > should return empty mentions array 1`] = `[]`;

exports[`unit:git/models/Commit > #mentions > should return non-empty mentions array 1`] = `
[
  "dependabot",
]
`;

exports[`unit:git/models/Commit > #pr > should return null 1`] = `null`;

exports[`unit:git/models/Commit > #pr > should return pull request number 1`] = `4`;

exports[`unit:git/models/Commit > #references > should return empty reference array 1`] = `[]`;

exports[`unit:git/models/Commit > #references > should return non-empty reference array 1`] = `
[
  {
    "action": null,
    "number": 15358,
    "owner": "prettier",
    "ref": "#15358",
    "repo": "prettier",
  },
  {
    "action": null,
    "number": 5715,
    "owner": "prettier",
    "ref": "#5715",
    "repo": "prettier",
  },
  {
    "action": null,
    "number": 11881,
    "owner": "prettier",
    "ref": "#11881",
    "repo": "prettier",
  },
  {
    "action": null,
    "number": 736,
    "owner": "dprint",
    "ref": "#736",
    "repo": "dprint",
  },
  {
    "action": null,
    "number": 432,
    "owner": "dprint",
    "ref": "#432",
    "repo": "dprint-plugin-typescript",
  },
]
`;

exports[`unit:git/models/Commit > #scope > should return commit scope 1`] = `"models"`;

exports[`unit:git/models/Commit > #scope > should return null 1`] = `null`;

exports[`unit:git/models/Commit > #sha > should return commit sha 1`] = `"e00383628638335e729d1cad2971b44ba03d3074"`;

exports[`unit:git/models/Commit > #subject > should return commit subject 1`] = `"\`Version\`"`;

exports[`unit:git/models/Commit > #tags > should return empty tags array 1`] = `[]`;

exports[`unit:git/models/Commit > #tags > should return non-empty tags array 1`] = `
[
  "grease@2.0.0",
]
`;

exports[`unit:git/models/Commit > #toJSON > should return json-serializable commit 1`] = `
{
  "author": {
    "email": "unicornware@flexdevelopment.llc",
    "name": "Lexus Drumgold",
  },
  "body": "",
  "breaking": false,
  "breaks": [],
  "date": "2021-09-18T23:29:18-04:00",
  "hash": "e003836",
  "header": "release: @flex-development/grease@2.0.0",
  "mentions": [],
  "pr": null,
  "references": [],
  "scope": null,
  "sha": "e00383628638335e729d1cad2971b44ba03d3074",
  "subject": "@flex-development/grease@2.0.0",
  "tags": [
    "grease@2.0.0",
  ],
  "trailers": [],
  "type": "release",
  "version": "grease@2.0.0",
}
`;

exports[`unit:git/models/Commit > #trailers > should return empty trailers array 1`] = `[]`;

exports[`unit:git/models/Commit > #trailers > should return non-empty trailers array 1`] = `
[
  {
    "token": "Signed-off-by",
    "value": "dependabot[bot] <support@github.com>",
  },
  {
    "token": "Co-authored-by",
    "value": "dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>",
  },
]
`;

exports[`unit:git/models/Commit > #type > should return empty commit type 1`] = `""`;

exports[`unit:git/models/Commit > #type > should return non-empty commit type 1`] = `"build"`;

exports[`unit:git/models/Commit > #version > should return null 1`] = `null`;

exports[`unit:git/models/Commit > #version > should return release tag 1`] = `"grease@2.0.0"`;
